File: server/src/initialize.ts
import { RequestMessage } from "./server";


type ServerCapabilities = Record<string, unknown>
interface InitializeResult {
    capabilities: ServerCapabilities;
    serverInfo?: {
        name: string;
        version?: string;
    };
}

export function initialize(message: RequestMessage): InitializeResult {
    return {
        capabilities: {
            completionProvider: {},
            textDocumentSync: 1
        },
        serverInfo: {
            name: "lsp-from-scratch",
            version: "0.0.1"
        }
    }
}

File: server-go/pkg/lsp/initialize.go
package lsp

type InitializeRequest struct {
	Request
	Params InitializeRequestParams `json:"params"`
}

type InitializeRequestParams struct {
	ClientInfo *ClientInfo `json:"clientInfo"`
}

type ClientInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

type InitializeResponse struct {
	Response
	Result InitializeResult `json:"result"`
}

type InitializeResult struct {
	Capabilities ServerCapabilities `json:"capabilities"`
	ServerInfo   ServerInfo         `json:"serverInfo"`
}

type ServerCapabilities struct {
	TextDocumentSync   int                `json:"textDocumentSync"`
	PositionEncoding   string             `json:"positionEncoding,omitempty"`
	CompletionProvider CompletionProvider `json:"completionProvider,omitempty"`
}

type ServerInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

type CompletionProvider struct {
}

func NewInitializeResponse(id int) InitializeResponse {
	return InitializeResponse{
		Response: Response{
			ID:  &id,
			RPC: "2.0",
		},
		Result: InitializeResult{
			Capabilities: ServerCapabilities{
				TextDocumentSync: 1,
				PositionEncoding: "utf-16",
			},
			ServerInfo: ServerInfo{
				Name:    "lsp-from-scratch",
				Version: "0.0.1",
			},
		},
	}
}
